### PR TITLE
[TableGen] Allow emitter callbacks to use `const RecordKeeper &`

### PR DESCRIPTION
**Pull request by jurahul**
- Refactor TableGen backend options to allow specifying a callback
  function that takes either a const reference or a non-const reference
  to `RecordKeeper`. This will enable gradual migration of code to use
  const references and pointers to `RecordKeeper` and  `Record` in the
  TableGen backends. 

  - Refactor handling of the callback command line options. Move variable
  for the command line option from the header to the CPP file, and add a
  function `ApplyCallback` to apply the selected callback.

  - Change callbacks in TableGen.cpp to take const reference. They use the
  `Opt` class to register their callbacks.

  - Change IntrinsicEmitter to use the `OptClass` to define its callbacks.
  It already uses a const reference in the implementation.

### CODE REVIEW

File: llvm/utils/TableGen/TableGen.cpp [hunk 1]
```diff
@@ -33,50 +33,47 @@ cl::opt<bool> EmitLongStrLiterals(
     cl::Hidden, cl::init(true));
 } // end namespace llvm
 
-namespace {
+static cl::OptionCategory PrintEnumsCat("Options for -print-enums");
```
Comment (mshockwave) [inline llvm/utils/TableGen/TableGen.cpp:36]: nit: I might be pedantic but I would prefer to split this anonymous namespace -> static refactoring to a separate commit / PR.

File: llvm/include/llvm/TableGen/TableGenBackend.h [hunk 2]
```diff
@@ -24,29 +22,38 @@ class RecordKeeper;
...
-using FnT = void (*)(RecordKeeper &Records, raw_ostream &OS);
-
-struct OptCreatorT {
-  static void *call();
-};
-
-extern ManagedStatic<cl::opt<FnT>, OptCreatorT> Action;
+// Support const and non-const forms of callback functions.
+using FnNonConstT = void (*)(RecordKeeper &Records, raw_ostream &OS);
+using FnConstT = void (*)(const RecordKeeper &Records, raw_ostream &OS);
```
Comment (Pierre-vh) [inline llvm/include/llvm/TableGen/TableGenBackend.h:27]: function_ref?

File: llvm/lib/TableGen/TableGenBackend.cpp [hunk 3]
```diff
@@ -12,22 +12,76 @@
...
 using namespace llvm;
+using namespace TableGen::Emitter;
 
 const size_t MAX_LINE_LEN = 80U;
 
-namespace llvm::TableGen::Emitter {
-ManagedStatic<cl::opt<FnT>, OptCreatorT> Action;
-void *OptCreatorT::call() {
-  return new cl::opt<FnT>(cl::desc("Action to perform:"));
+using FnT = std::variant<FnNonConstT, FnConstT>;
```
Comment (Pierre-vh) [inline llvm/lib/TableGen/TableGenBackend.cpp:29]: nit: I'm not a fan of using variant here, could we simplify this by just storing non-const version all the time ? Cast the const away and always call it with the mutable version. Then this code doesn't care whether the actual target uses a const or non const RK

Comment (jurahul) [reply llvm/lib/TableGen/TableGenBackend.cpp:29]: Is that considered safe/kosher per LLVM coding standards?

File: llvm/utils/TableGen/TableGen.cpp (hunk 1 above)
Comment (jurahul) [reply llvm/utils/TableGen/TableGen.cpp:36]: Done, I started https://github.com/llvm/llvm-project/pull/105504. I'll update this diff once that gets merged.

File: llvm/lib/TableGen/TableGenBackend.cpp (hunk 3 above)
Comment (jurahul) [reply llvm/lib/TableGen/TableGenBackend.cpp:29]: Also, what are the concerns w.r.t variant<>?  I do see variant<> being use in LLVM code, including TableGen, though admittedly not as widely.

File: llvm/utils/TableGen/TableGen.cpp (hunk 1 above)
Comment (jurahul) [reply llvm/utils/TableGen/TableGen.cpp:36]: This is committed now.

File: llvm/include/llvm/TableGen/TableGenBackend.h (hunk 2 above)
Comment (jurahul) [reply llvm/include/llvm/TableGen/TableGenBackend.h:27]: That does not seem to work, in the sense function_ref<> seems to erase the const vs non-const distinction. To verify, I added the following test to FunctionRefTest.cpp

```C++

std::string returns(function_ref<void(int &)> F) { return "int&"; }
std::string returns(function_ref<void(const int &)> F) { return "const int&"; }

TEST(FunctionRefTest, SFINAE) {
  EXPECT_EQ("int&", returns([](int&) { }));
  EXPECT_EQ("string", returns([](const int&) { }));
}
```

and I get the following error:

```
/home/rjoshi/upstream_llvm/llvm-project/llvm/unittests/ADT/FunctionRefTest.cpp:63:23: error: call to 'returns' is ambiguous
  EXPECT_EQ("string", returns([](const int&) { }));
/home/rjoshi/upstream_llvm/llvm-project/llvm/unittests/ADT/FunctionRefTest.cpp:55:13: note: candidate function
std::string returns(function_ref<void(int &)> F) { return "int&"; }
            ^
/home/rjoshi/upstream_llvm/llvm-project/llvm/unittests/ADT/FunctionRefTest.cpp:56:13: note: candidate function
std::string returns(function_ref<void(const int &)> F) { return "const int&"; }
```

Comment (jurahul) [reply llvm/include/llvm/TableGen/TableGenBackend.h:27]: This can actually be used to our advantage for eliminating the std::variant.

File: llvm/lib/TableGen/TableGenBackend.cpp (hunk 3 above)
Comment (Pierre-vh) [reply llvm/lib/TableGen/TableGenBackend.cpp]: Note that this is just a nitpick, not something I'd block the diff over and this is strictly my opinion: Variant is great in some cases, but it's also a complicated class with unclear performance implications. I prefer to avoid it when possible.

In this case, there is another solution that replaces the variant with just a trivial pointer.

> Is that considered safe/kosher per LLVM coding standards?

I don't see why it wouldn't be, I would be concerned for safety if you were casting the non-const version to the const version, then the call site could assume RecordKeeper isn't modified when it is. The other way around is fine -the call site assumes we modify RK when we actually don't, just like in any other function called with a non-const ref

Feel free to wait for other reviewers to weigh in, if someone has a strong preference towards one or the other then just listen to that instead :)

Comment (jurahul) [reply llvm/lib/TableGen/TableGenBackend.cpp]: As I said above, I was actually able to get rid of it, and it simplified the code as well.

File: llvm/lib/TableGen/TableGenBackend.cpp [hunk 4]
```diff
@@ -12,22 +12,65 @@
...
+
+  OptionValue(const FnT &V) { this->setValue(V); }
+
+  OptionValue<FnT> &operator=(const FnT &V) {
+    setValue(V);
+    return *this;
+  }
+
+private:
+  void anchor() override {}
```
Comment (Pierre-vh) [inline llvm/lib/TableGen/TableGenBackend.cpp:45]: I don't think you need an anchor if the class is contained inside a .cpp file, it's just useful in headers IIRC

Comment (Pierre-vh) [inline llvm/lib/TableGen/TableGenBackend.cpp:45]: I don't think you need an anchor if the class is contained inside a .cpp file, it's just useful in headers IIRC

File: llvm/unittests/ADT/FunctionRefTest.cpp [hunk 5]
```diff
@@ -59,4 +59,12 @@ TEST(FunctionRefTest, SFINAE) {
   EXPECT_EQ("string", returns([] { return "hello"; }));
 }
 
+TEST(FunctionRefTest, Equality) {
+  function_ref<int()> X = [] { return 1; };
+  function_ref<int()> Y = [] { return 2; };
+  EXPECT_TRUE(!(X == Y));
```
Comment (Pierre-vh) [inline llvm/unittests/ADT/FunctionRefTest.cpp:65]:
```suggestion
  EXPECT_FALSE(X == Y);
  EXPECT_TRUE(X != Y);
```

also test something like
```
const auto Lambda = []() { return 0 };
function_ref<int()> A(Lambda), B (Lambda);
EXPECT_TRUE(A == B);
EXPECT_FALSE(A != B);
```

I think != should be automatically available but if it isn't then it's fine, don't bother testing it
```

File: llvm/include/llvm/ADT/STLFunctionalExtras.h [hunk 6]
```diff
@@ -69,6 +69,10 @@ class function_ref<Ret(Params...)> {
   }
 
   explicit operator bool() const { return callback; }
+
+  bool operator==(const function_ref<Ret(Params...)> &Other) const {
```
Comment (Pierre-vh) [inline llvm/include/llvm/ADT/STLFunctionalExtras.h:73]: Why is this needed now?

File: llvm/lib/TableGen/TableGenBackend.cpp (hunk 4 above)
Comment (jurahul) [reply llvm/lib/TableGen/TableGenBackend.cpp:45]: Agreed, will remove.

Comment (jurahul) [reply llvm/lib/TableGen/TableGenBackend.cpp:45]: Duplicate comment, resolving.

File: llvm/include/llvm/ADT/STLFunctionalExtras.h (hunk 6 above)
Comment (jurahul) [reply llvm/include/llvm/ADT/STLFunctionalExtras.h:73]: This is needed now because of its use in `cl::OptionValueCopy<FnT>`, which uses the `==` operator.

File: llvm/unittests/ADT/FunctionRefTest.cpp (hunk 5 above)
Comment (jurahul) [reply llvm/unittests/ADT/FunctionRefTest.cpp]: The != is not synthesized from ==, so dropped those test and added the one you suggested.

File: llvm/lib/TableGen/TableGenBackend.cpp (hunk 4 above)
Comment (jurahul) [reply llvm/lib/TableGen/TableGenBackend.cpp]: Done.

**Review by Pierre-vh — APPROVED** [review]
LGTM but please wait a working day or two to see if anyone else has concerns

File: llvm/include/llvm/TableGen/TableGenBackend.h (hunk 2 above)
Comment (jurahul) [reply llvm/include/llvm/TableGen/TableGenBackend.h]: Resolving this thread.

File: llvm/lib/TableGen/TableGenBackend.cpp (hunk 3 above)
Comment (jurahul) [reply llvm/lib/TableGen/TableGenBackend.cpp]: Resolving.

**Comment by jurahul** [comment]
> LGTM but please wait a working day or two to see if anyone else has concerns

Awesome, yeah I'll wait till next week to merge it given the weekend is approaching.
